File: 알고리즘/전승민/skillTree/src/SkillTree.java
public class SkillTree {
    public static void main(String[] args) {
        String skill = "CBD";
        String[] skill_trees = new String[]{"BACDE", "CBADF", "AECB", "BDA"};
        System.out.println(new SkillTree().solution(skill, skill_trees));
    }

    public int solution(String skill, String[] skill_trees) {
        int result = 0;
        int[] alphabetArr = checkSkillSequence(skill);

        for(String s : skill_trees) {
            if(isCollectSkillTree(s, alphabetArr)) {
                result++;
            }
        }

        return result;
    }

    private static int[] checkSkillSequence(String skill) {
        int[] alphabetArr = new int[26];
        int length = skill.length();
        for(int i = 0; i < length; i++) {
            alphabetArr[getAlphabetIndex(skill.charAt(i))] = i+1;
        }
        return alphabetArr;
    }

    private static int getAlphabetIndex(char c) {
        return c - 65;
    }

    private static boolean isCollectSkillTree(String skillTree, int[] alphabetArr) {
        int length = skillTree.length();
        int count = 1;
        for(int i=0; i<length; i++) {
            int alphabet = getAlphabetIndex(skillTree.charAt(i));
            if(alphabetArr[alphabet]==0) { }
            else if(alphabetArr[alphabet]==count) {
                count++;
            } else {
                count = -1;
                break;
            }
        }
        if(count!=-1) {
            return true;
        } else {
            return false;
        }
    }
}

File: 알고리즘/전승민/skillTree/src/SkillTree.kt
fun skillTree(skill: String, skill_trees: Array<String>): Int
        = with(checkSkillSequence(skill)) {
    var result = 0
    for (s: String in skill_trees) {
        if (isCollectSkillTree(this, s)) result++
    }
    result
}

fun checkSkillSequence(skill: String): Array<Int?> =
        arrayOfNulls<Int>(26).apply {
            var count = 1
            skill.forEach {
                this[it.index] = count++
            }
        }

fun isCollectSkillTree(alphabetArr: Array<Int?>, skill: String): Boolean {
    var count = 1
    skill.forEach {
        if (alphabetArr[it.index] ?: 0 > count) {
            return false
        } else if(alphabetArr[it.index] ?: 0 == count) {
            count++
        }
    }
    return true
}

//https://programmers.co.kr/learn/courses/30/lessons/49993
